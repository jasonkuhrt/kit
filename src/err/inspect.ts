import { Arr } from '#arr/index.js'
import { Language } from '#language/index.js'
import { Obj } from '#obj/index.js'
import { Rec } from '#rec/index.js'
import { Str } from '#str/index.js'
import { Char } from '#str/str.js'
import type { Ts } from '#ts/index.js'
import { cyan, red } from 'ansis'
import { cleanStack } from './stack.js'
import { is } from './type.js'
import type { Context } from './types.js'

interface EnvironmentConfigurableOptionSpec<$Name extends string = string, $Type = any> {
  name: $Name
  envVarNamePrefix: string
  default: NoInfer<$Type>
  description?: string
  parse: (envVarValue: string) => $Type
}

const makeEnvVarName = (spec: EnvironmentConfigurableOptionSpec) => {
  return Str.Case.upper(
    Str.Case.snake(`${spec.envVarNamePrefix}_${spec.name}`),
  )
}

export type InferOptions<$EnvironmentConfigurableOptions extends EnvironmentConfigurableOptionSpec[]> = Ts.Simplify<
  Arr.ReduceWithIntersection<_InferOptions<$EnvironmentConfigurableOptions>>
>

export type _InferOptions<$EnvironmentConfigurableOptions extends EnvironmentConfigurableOptionSpec[]> = {
  [i in keyof $EnvironmentConfigurableOptions]: {
    [_ in $EnvironmentConfigurableOptions[i]['name']]?: ReturnType<$EnvironmentConfigurableOptions[i]['parse']>
  }
}

const define = <const options extends EnvironmentConfigurableOptionSpec[]>(options: options): options => {
  return options
}

interface EnvironmentConfigurableOptionInput<$Spec extends EnvironmentConfigurableOptionSpec> {
  spec: $Spec
  value: any
  source: 'default' | 'environment'
}

type Resolve<$Specs extends EnvironmentConfigurableOptionSpec[]> = Ts.Simplify<
  Arr.ReduceWithIntersection<_Resovle<$Specs>>
>

type _Resovle<$Specs extends EnvironmentConfigurableOptionSpec[]> = {
  [i in keyof $Specs]: {
    [_ in $Specs[i]['name']]: {
      spec: $Specs[i]
      value: ReturnType<$Specs[i]['parse']>
      source: 'default' | 'environment'
    }
  }
}

const resolve = <const specs extends EnvironmentConfigurableOptionSpec[]>(
  specs: specs,
  input: InferOptions<specs>,
): Resolve<specs> => {
  const config = Rec.create<EnvironmentConfigurableOptionInput<specs[number]>>()
  const input$ = input as Record<string, any>

  for (const spec of specs) {
    const processValue = Language.process.env[makeEnvVarName(spec)]
    if (processValue !== undefined) {
      config[spec.name] = {
        spec,
        value: spec.parse(processValue),
        source: 'environment',
      }
      continue
    }
    if (spec.name in input && input$[spec.name] !== undefined) {
      config[spec.name] = {
        spec,
        value: input$[spec.name],
        source: 'default',
      }
      continue
    }
    config[spec.name] = {
      spec,
      value: spec.default,
      source: 'default',
    }
  }

  return config as any
}

// ---------------

const optionSpecs = define([
  {
    name: 'color',
    envVarNamePrefix: 'errorDsiplay',
    description: 'Should output be colored for easier reading',
    default: true,
    parse: (envVarValue) => envVarValue === '0' || envVarValue === 'false' ? false : true,
  },
  {
    name: 'stackTraceColumns',
    envVarNamePrefix: 'errorDsiplay',
    description: 'The column count to display before truncation begins',
    default: 120,
    parse: (envVarValue) => parseInt(envVarValue, 10),
  },
  {
    name: 'identColumns',
    envVarNamePrefix: 'errorDsiplay',
    description: 'The column count to use for indentation',
    default: 4,
    parse: (envVarValue) => parseInt(envVarValue, 10),
  },
])

export type InspectOptions = InferOptions<typeof optionSpecs>

export type InspectConfig = Resolve<typeof optionSpecs>

export const formatTitle = (indent: string, text: string, config: InspectConfig) => {
  const title = `\n${indent}${text.toUpperCase()}\n`
  return config.color.value ? cyan(title) : title
}

/**
 * Render an error to a string with nice formatting including causes, aggregate errors, context, and stack traces.
 */

export const inspect = (error: Error, options?: InspectOptions): string => {
  const config = resolve(optionSpecs, options ?? {})

  let inspection = _inspectResursively(error, '', config)

  inspection += `\n${formatTitle('', 'Environment Variable Formatting', config)}`

  // todo: indent should not indent empty lines
  inspection += Str.indent(
    `\nYou can control formatting of this error display with the following environment variables.\n`,
    config.identColumns.value,
  )

  inspection += '\n'

  for (const [_, state] of Obj.entries(config)) {
    inspection += Str.indent(
      `${makeEnvVarName(state.spec)} â€“ The column count to display before truncation ${
        state.source === 'environment'
          ? `(currently set to ${state.value})`
          : `(currently unset, defaulting to ${state.value})`
      }`,
      config.identColumns.value,
    )
    inspection += '\n'
  }

  return inspection
}

const _inspectResursively = (error: Error, parentIndent: string, config: InspectConfig): string => {
  const formatIndent = Char.spaceRegular.repeat(config.identColumns.value)
  if (!is(error)) {
    return parentIndent + String(error)
  }

  const lines: string[] = []

  // Handle AggregateError
  if (error instanceof AggregateError && error.errors.length > 0) {
    lines.push(formatTitle(parentIndent, 'errors', config))
    lines.push(...error.errors.reduce<string[]>((acc, err, index) => {
      acc.push(formatTitle(parentIndent, `[${index}]`, config))
      acc.push(_inspectResursively(err, parentIndent + formatIndent, config))
      return acc
    }, []))
    return Str.unlines(lines)
  }

  // Render the main error message
  const errorName = config.color.value ? red(error.name) : error.name
  lines.push(`${parentIndent}${errorName}`)
  lines.push('')
  lines.push(`${parentIndent}${formatIndent}${error.message}`)

  // Handle cause if present
  if ('cause' in error && error.cause instanceof Error) {
    lines.push(formatTitle(parentIndent, 'cause', config))
    lines.push(_inspectResursively(error.cause, parentIndent + formatIndent, config))
  }

  // Handle context property if present
  if ('context' in error && error.context !== undefined) {
    lines.push(formatTitle(parentIndent, 'context', config))
    // todo: pretty object rendering
    lines.push(String((error as Error & { context: Context }).context))
  }

  // Handle stack property if present
  if (error.stack) {
    // Clean the stack trace first
    const cleanedStack = cleanStack(error.stack, {
      removeInternal: true,
      maxFrames: 15,
      filterPatterns: ['node_modules', 'node:internal'],
    })

    const stack = Str.unlines(
      Str
        .lines(cleanedStack)
        // Stacks include the message by default, we already showed that above.
        .slice(1)
        .map(_ => Str.truncate(`${parentIndent}${formatIndent}${_.trim()}`, config.stackTraceColumns.value)),
    )
    lines.push(formatTitle(parentIndent, 'stack', config))
    lines.push(stack)
  }

  return Str.unlines(lines)
}
