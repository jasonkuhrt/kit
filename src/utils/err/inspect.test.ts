import { describe, expect, test } from 'vitest'
import { inspect } from './inspect.js'

describe('inspect', () => {
  describe('basic error rendering', () => {
    test('simple error', () => {
      const error = new Error('Something went wrong')
      error.stack = `Error: Something went wrong
    at Function.execute (/app/src/index.ts:10:15)
    at process (/app/src/main.ts:20:10)
    at Object.<anonymous> (/app/index.js:5:1)`

      const result = inspect(error, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
    })

    test('error with no stack', () => {
      const error = new Error('No stack available')
      delete error.stack
      const result = inspect(error, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
    })

    test('error with empty message', () => {
      const error = new Error('')
      error.stack = `Error
    at Function.execute (/app/src/index.ts:10:15)`
      const result = inspect(error, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
    })
  })

  describe('error with context', () => {
    test('simple context object', () => {
      const error = new Error('API failed')
      ;(error as any).context = { userId: 123, endpoint: '/api/users' }
      error.stack = `Error: API failed
    at apiCall (/app/src/api.ts:45:10)
    at handleRequest (/app/src/server.ts:100:15)`

      const result = inspect(error, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
    })

    test('nested context object', () => {
      const error = new Error('Complex failure')
      ;(error as any).context = {
        request: {
          method: 'POST',
          url: '/api/data',
          headers: { 'content-type': 'application/json' },
        },
        timestamp: '2024-01-15T10:30:00Z',
        traceId: '123-456-789',
      }
      error.stack = `Error: Complex failure
    at process (/app/src/handler.ts:30:8)`

      const result = inspect(error, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
    })
  })

  describe('error with cause', () => {
    test('single level cause', () => {
      const cause = new Error('Database connection failed')
      cause.stack = `Error: Database connection failed
    at connect (/app/src/db.ts:15:10)`

      const error = new Error('Failed to fetch user', { cause })
      error.stack = `Error: Failed to fetch user
    at fetchUser (/app/src/user.ts:30:15)
    at handleRequest (/app/src/api.ts:50:20)`

      const result = inspect(error, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
    })

    test('nested causes', () => {
      const rootCause = new Error('Network timeout')
      rootCause.stack = `Error: Network timeout
    at socket (/app/src/net.ts:5:5)`

      const middleCause = new Error('Redis connection failed', { cause: rootCause })
      middleCause.stack = `Error: Redis connection failed
    at redis.connect (/app/src/cache.ts:10:10)`

      const error = new Error('Session creation failed', { cause: middleCause })
      error.stack = `Error: Session creation failed
    at createSession (/app/src/session.ts:25:15)`

      const result = inspect(error, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
    })
  })

  describe('AggregateError', () => {
    test('simple aggregate error', () => {
      const error1 = new Error('Service A failed')
      error1.stack = `Error: Service A failed
    at serviceA (/app/src/services/a.ts:10:5)`

      const error2 = new Error('Service B timeout')
      error2.stack = `Error: Service B timeout
    at serviceB (/app/src/services/b.ts:20:10)`

      const aggregate = new AggregateError([error1, error2], 'Multiple services failed')
      aggregate.stack = `AggregateError: Multiple services failed
    at coordinator (/app/src/coordinator.ts:50:15)`

      const result = inspect(aggregate, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
    })

    test('aggregate error with context', () => {
      const error1 = new Error('Database error')
      error1.stack = `Error: Database error
    at query (/app/src/db.ts:30:10)`
      ;(error1 as any).context = { table: 'users', operation: 'INSERT' }

      const error2 = new Error('Cache error')
      error2.stack = `Error: Cache error
    at cache.set (/app/src/cache.ts:15:5)`
      ;(error2 as any).context = { key: 'user:123', ttl: 3600 }

      const aggregate = new AggregateError([error1, error2], 'Data layer failures')
      ;(aggregate as any).context = { transactionId: 'tx-456', timestamp: 1234567890 }
      aggregate.stack = `AggregateError: Data layer failures
    at transaction (/app/src/transaction.ts:100:20)`

      const result = inspect(aggregate, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
    })

    test('nested aggregate errors', () => {
      const innerError = new Error('Inner failure')
      innerError.stack = `Error: Inner failure
    at inner (/app/src/inner.ts:5:5)`

      const innerAggregate = new AggregateError([innerError], 'Inner aggregate')
      innerAggregate.stack = `AggregateError: Inner aggregate
    at processInner (/app/src/process.ts:20:10)`

      const outerError = new Error('Outer failure')
      outerError.stack = `Error: Outer failure
    at outer (/app/src/outer.ts:10:10)`

      const outerAggregate = new AggregateError([innerAggregate, outerError], 'Outer aggregate')
      outerAggregate.stack = `AggregateError: Outer aggregate
    at main (/app/src/main.ts:50:15)`

      const result = inspect(outerAggregate, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
    })
  })

  describe('stack trace filtering', () => {
    test('filters node_modules frames', () => {
      const error = new Error('Test error')
      error.stack = `Error: Test error
    at userCode (/app/src/user.ts:10:5)
    at node_modules/lib/index.js:100:20
    at node_modules/other/file.js:50:10
    at moreUserCode (/app/src/main.ts:30:15)`

      const result = inspect(error, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
      expect(result).toContain('elided')
      expect(result).not.toContain('node_modules/lib')
    })

    test('shows all frames filtered message', () => {
      const error = new Error('All filtered')
      error.stack = `Error: All filtered
    at node_modules/lib1/index.js:10:5
    at node_modules/lib2/index.js:20:10
    at node_modules/lib3/index.js:30:15`

      const result = inspect(error, { showHelp: false, color: false })
      expect(result).toMatchSnapshot()
      expect(result).toContain('all 3 frames elided')
    })
  })

  // Table-driven tests for comprehensive scenarios
  describe('comprehensive error scenarios', () => {
    // dprint-ignore
    const cases: {
      name: string
      createError: () => Error
      options?: Parameters<typeof inspect>[1]
      expectations?: {
        contains?: string[]
        notContains?: string[]
      }
    }[] = [
    {
      name: 'custom error class',
      createError: () => {
        class CustomError extends Error {
          constructor(message: string, public code: string) {
            super(message)
            this.name = 'CustomError'
          }
        }
        const error = new CustomError('Custom failure', 'ERR_CUSTOM')
        error.stack = `CustomError: Custom failure\n    at test (/app/test.ts:10:5)`
        return error
      },
      options: { showHelp: false, color: false },
      expectations: {
        contains: ['CustomError', 'Custom failure'],
      },
    },
    {
      name: 'error with very long message',
      createError: () => {
        const longMessage = 'This is a very long error message that '.repeat(10) + 'ends here'
        const error = new Error(longMessage)
        error.stack = `Error: ${longMessage}\n    at test (/app/test.ts:10:5)`
        return error
      },
      options: { showHelp: false, color: false, stackTraceColumns: 80 },
      expectations: {
        contains: ['very long error message'],
      },
    },
    {
      name: 'error with special characters in context',
      createError: () => {
        const error = new Error('Special chars')
        ;(error as any).context = {
          'special-key': 'value',
          'key with spaces': true,
          'emojiðŸŽ‰': 'test',
          nested: { 'more-special': 123 },
        }
        error.stack = `Error: Special chars\n    at test (/app/test.ts:10:5)`
        return error
      },
      options: { showHelp: false, color: false },
      expectations: {
        contains: ['special-key', 'key with spaces', 'emojiðŸŽ‰'],
      },
    },
    {
      name: 'error with circular reference in context',
      createError: () => {
        const error = new Error('Circular ref')
        const circular: any = { a: 1 }
        circular.self = circular
        ;(error as any).context = circular
        error.stack = `Error: Circular ref\n    at test (/app/test.ts:10:5)`
        return error
      },
      options: { showHelp: false, color: false },
      expectations: {
        contains: ['Circular ref'],
      },
    },
    ]

    for (const { name, createError, options, expectations } of cases) {
      test(name, () => {
        const error = createError()
        const result = inspect(error, options)

        // Use snapshot for full output
        expect(result).toMatchSnapshot()

        // Additional assertions
        if (expectations?.contains) {
          for (const text of expectations.contains) {
            expect(result).toContain(text)
          }
        }
        if (expectations?.notContains) {
          for (const text of expectations.notContains) {
            expect(result).not.toContain(text)
          }
        }
      })
    }
  })

  describe('environment variable configuration', () => {
    test('shows help section by default', () => {
      const error = new Error('Test')
      error.stack = `Error: Test\n    at test (/app/test.ts:10:5)`

      const result = inspect(error, { color: false })
      expect(result).toContain('Environment Variable Configuration')
      expect(result).toContain('ERROR_DISPLAY_COLOR')
      expect(result).toContain('ERROR_DISPLAY_SHOW_HELP')
    })

    test('hides help section when disabled', () => {
      const error = new Error('Test')
      error.stack = `Error: Test\n    at test (/app/test.ts:10:5)`

      const result = inspect(error, { showHelp: false, color: false })
      expect(result).not.toContain('Environment Variable Configuration')
      expect(result).not.toContain('ERROR_DISPLAY_COLOR')
    })
  })

  describe('visual guides (with color)', () => {
    test('shows tree characters for nested errors', () => {
      const cause = new Error('Cause')
      cause.stack = `Error: Cause\n    at cause (/app/cause.ts:5:5)`

      const error = new Error('Main', { cause })
      error.stack = `Error: Main\n    at main (/app/main.ts:10:10)`

      // Note: We're not testing actual colors here, just structure
      const result = inspect(error, { showHelp: false, color: true })
      expect(result).toMatchSnapshot()
    })

    test('shows proper indentation for aggregate errors', () => {
      const errors = [
        new Error('First'),
        new Error('Second'),
        new Error('Third'),
      ]
      errors[0]!.stack = `Error: First\n    at first (/app/first.ts:1:1)`
      errors[1]!.stack = `Error: Second\n    at second (/app/second.ts:2:2)`
      errors[2]!.stack = `Error: Third\n    at third (/app/third.ts:3:3)`

      const aggregate = new AggregateError(errors, 'Multiple')
      aggregate.stack = `AggregateError: Multiple\n    at main (/app/main.ts:10:10)`

      const result = inspect(aggregate, { showHelp: false, color: true })
      expect(result).toMatchSnapshot()
    })
  })
})
