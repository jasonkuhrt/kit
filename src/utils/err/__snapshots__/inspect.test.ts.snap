// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`inspect > AggregateError > aggregate error with context 1`] = `
"AggregateError: Data layer failures
@ transaction /app/src/transaction.ts:100:20
{
  transactionId: 'tx-456',
  timestamp: 1234567890
}
    â†“
    â”‚
  0 â”œâ”€ Error: Database error
    â”‚  @ query /app/src/db.ts:30:10
    â”‚  {
    â”‚    table: 'users',
    â”‚    operation: 'INSERT'
    â”‚  }
    â”‚
  1 â”œâ”€ Error: Cache error
    â”‚  @ cache.set /app/src/cache.ts:15:5
    â”‚  {
    â”‚    key: 'user:123',
    â”‚    ttl: 3600
    â”‚  }
    â””"
`;

exports[`inspect > AggregateError > nested aggregate errors 1`] = `
"AggregateError: Outer aggregate
@ main /app/src/main.ts:50:15
    â†“
    â”‚
  0 â”œâ”€ AggregateError: Inner aggregate
    â”‚  @ processInner /app/src/process.ts:20:10
    â”‚  â†“
    â”‚    â””â”€ Error: Inner failure
    â”‚     @ inner /app/src/inner.ts:5:5
    â”‚
  1 â”œâ”€ Error: Outer failure
    â”‚  @ outer /app/src/outer.ts:10:10
    â””"
`;

exports[`inspect > AggregateError > simple aggregate error 1`] = `
"AggregateError: Multiple services failed
@ coordinator /app/src/coordinator.ts:50:15
    â†“
    â”‚
  0 â”œâ”€ Error: Service A failed
    â”‚  @ serviceA /app/src/services/a.ts:10:5
    â”‚
  1 â”œâ”€ Error: Service B timeout
    â”‚  @ serviceB /app/src/services/b.ts:20:10
    â””"
`;

exports[`inspect > basic error rendering > error with empty message 1`] = `
"Error
@ Function.execute /app/src/index.ts:10:15"
`;

exports[`inspect > basic error rendering > error with no stack 1`] = `"Error: No stack available"`;

exports[`inspect > basic error rendering > simple error 1`] = `
"Error: Something went wrong
@ Function.execute /app/src/index.ts:10:15
@ process          /app/src/main.ts:20:10
@ <?>              /app/index.js:5:1"
`;

exports[`inspect > comprehensive error scenarios > custom error class 1`] = `
"CustomError: Custom failure
@ test /app/test.ts:10:5"
`;

exports[`inspect > comprehensive error scenarios > error with circular reference in context 1`] = `
"Error: Circular ref
@ test /app/test.ts:10:5
{
  a: 1,
  self: [Circular]
}"
`;

exports[`inspect > comprehensive error scenarios > error with special characters in context 1`] = `
"Error: Special chars
@ test /app/test.ts:10:5
{
  'special-key': 'value',
  'key with spaces': true,
  'emojiðŸŽ‰': 'test',
  nested: {
    'more-special': 123
  }
}"
`;

exports[`inspect > comprehensive error scenarios > error with very long message 1`] = `
"Error: This is a very long error message that This is a very long error message that This is a very long error message that This is a very long error message that This is a very long error message that This is a very long error message that This is a very long error message that This is a very long error message that This is a very long error message that This is a very long error message that ends here
@ test /app/test.ts:10:5"
`;

exports[`inspect > error with cause > nested causes 1`] = `
"Error: Session creation failed
@ createSession /app/src/session.ts:25:15
  â†“
  Error: Redis connection failed
  @ redis.connect /app/src/cache.ts:10:10
    â†“
    Error: Network timeout
    @ socket /app/src/net.ts:5:5"
`;

exports[`inspect > error with cause > single level cause 1`] = `
"Error: Failed to fetch user
@ fetchUser     /app/src/user.ts:30:15
@ handleRequest /app/src/api.ts:50:20
  â†“
  Error: Database connection failed
  @ connect /app/src/db.ts:15:10"
`;

exports[`inspect > error with context > nested context object 1`] = `
"Error: Complex failure
@ process /app/src/handler.ts:30:8
{
  request: {
    method: 'POST',
    url: '/api/data',
    headers: {
      'content-type': 'application/json'
    }
  },
  timestamp: '2024-01-15T10:30:00Z',
  traceId: '123-456-789'
}"
`;

exports[`inspect > error with context > simple context object 1`] = `
"Error: API failed
@ apiCall       /app/src/api.ts:45:10
@ handleRequest /app/src/server.ts:100:15
{
  userId: 123,
  endpoint: '/api/users'
}"
`;

exports[`inspect > stack trace filtering > filters node_modules frames 1`] = `
"Error: Test error
@ userCode     /app/src/user.ts:10:5
@ moreUserCode /app/src/main.ts:30:15
@ ...          elided 2 frames"
`;

exports[`inspect > stack trace filtering > shows all frames filtered message 1`] = `
"Error: All filtered
@ ...             all 3 frames elided"
`;

exports[`inspect > visual guides (with color) > shows proper indentation for aggregate errors 1`] = `
"AggregateError: Multiple
@ main /app/main.ts:10:10
    â†“
    â”‚
  0 â”œâ”€ Error: First
    â”‚  @ first /app/first.ts:1:1
    â”‚
  1 â”œâ”€ Error: Second
    â”‚  @ second /app/second.ts:2:2
    â”‚
  2 â”œâ”€ Error: Third
    â”‚  @ third /app/third.ts:3:3
    â””"
`;

exports[`inspect > visual guides (with color) > shows tree characters for nested errors 1`] = `
"Error: Main
@ main /app/main.ts:10:10
  â†“
  Error: Cause
  @ cause /app/cause.ts:5:5"
`;
